### ISSUE TITLE
Infinite Loop in jemalloc 4.2.0

### ISSUE THREAD

**Issue by cferris1000** [jemalloc/jemalloc#390]
The newly released version of jemalloc can cause an infinite loop which leads to crashes when the stack overflows.

The circumstances that cause the crash:

- TLS is not being used to handle thread local variables. Instead, the code is using pthread primitives.
- A new thread is started, but the program has been running for a while with other threads doing allocations.
- The new thread calls into an allocation routine, which eventually tries to allocate the memory needed for  thread local variable data.
- Another thread winds up setting runcur to NULL for the bin in which the new thread is about to allocate.
- In the function arena_malloc_small, since runcur is NULL, a call to arena_bin_malloc_hard is made instead of arena_run_reg_alloc. The arena_bin_malloc_hard call eventually leads to a call to get the tsd, which tries to allocate the memory again and infinite loop.

To verify this, I took the jemalloc 4.2.0 release, built it with JEMALLOC_TLS undefined in include/jemalloc/internal/jemalloc_internal_defs.h, and changed the number of arenas to 1 to make it more likely to trip the problem. I built this on a ubuntu derived version of linux, using --with-jemalloc_prefix=je_. I did the same with 4.1.0 and 4.1.1, but neither exhibits this problem.

I compiled my small test program like so:

g++ -o free-4.2.0 -std=c++11 free.cpp -Ijemalloc-4.2.0/include jemalloc-4.2.0/lib/libjemalloc.a -lpthread

The test code is:

```
#include <stdio.h>

#include <thread>
#include <vector>

#include <jemalloc/jemalloc.h>

int main() {
  for (size_t i = 0; i < 100; i++) {
    std::thread thread([](){
      for (size_t k = 0; k < 10000; k++) {
        std::vector<void*> vector;
        for (size_t j = 0; j < 10000; j++) {
          vector.push_back(je_malloc(120));
        }
        for (auto ptr : vector) {
          je_free(ptr);
        }
      }
    });
    thread.detach();
  }
  for (size_t i = 0; i < 1000; i++) {
    std::thread thread([](){
      void* ptr = je_malloc(120);
      je_free(ptr);
    });
    thread.join();
  }
}
```

**Comment by glandium** [comment]
FWIW, I suspect something similar is happening on Windows. We have stack overflow crashes in jemalloc when running some Firefox unit tests with jemalloc 4.2.0.

**Comment by jasone** [comment]
cc289f40b6e7b119566abf8a1c09c97e08597a3d fixes this for the dev branch; I'll develop a similar fix for 4.2.1.

**Comment by jasone** [comment]
Fixed in 4.2.1: 09d7bdb314ec4d6a8618d6958809b540f8d3cff6.  Thanks for the detailed report; it made this quick work.
